cpp/src/hash/md5_hash.cu

```diff
@@ -27,46 +31,265 @@
+        break;
+    }
+
+    uint32_t buffer_element_as_int;
+    memcpy(&buffer_element_as_int, hash_state->buffer + g * 4, 4);
+    F = F + A + md5_hash_constants[j] + buffer_element_as_int;
+    A = D;
+    D = C;
+    C = B;
+    B = B + __funnelshift_l(F, F, md5_shift_constants[((j / 16) * 4) + (j % 4)]);
+  }
+
+  hash_state->hash_value[0] += A;
+  hash_state->hash_value[1] += B;
+  hash_state->hash_value[2] += C;
+  hash_state->hash_value[3] += D;
+
+  hash_state->buffer_length = 0;
+}
+
+/**
+ * @brief Core MD5 element processing function
+ *
+ * This accepts arbitrary data, handles it as bytes, and calls the hash step
+ * when the buffer is filled up to message_chunk_size bytes.
+ */
+void CUDA_DEVICE_CALLABLE md5_process_bytes(uint8_t const* data,
+                                            uint32_t len,
+                                            md5_intermediate_data* hash_state)
+{
+  hash_state->message_length += len;
+
+  // 64 bytes are processed in each hash step
+  uint32_t constexpr md5_chunk_size = 64;
+
+  if (hash_state->buffer_length + len < md5_chunk_size) {
+    // The buffer will not be filled by this data. We copy the new data into
+    // the buffer but do not trigger a hash step yet.
+    memcpy(hash_state->buffer + hash_state->buffer_length, data, len);
+    hash_state->buffer_length += len;
+  } else {
+    // The buffer will be filled by this data. Copy a chunk of the data to fill
+    // the buffer and trigger a hash step.
+    uint32_t copylen = md5_chunk_size - hash_state->buffer_length;
+    memcpy(hash_state->buffer + hash_state->buffer_length, data, copylen);
+    md5_hash_step(hash_state);
+
+    // Take buffer-sized chunks of the data and do a hash step on each chunk.
+    while (len > md5_chunk_size + copylen) {
+      memcpy(hash_state->buffer, data + copylen, md5_chunk_size);
+      md5_hash_step(hash_state);
+      copylen += md5_chunk_size;
+    }
+
+    // The remaining data chunk does not fill the buffer. We copy the data into
+    // the buffer but do not trigger a hash step yet.
+    memcpy(hash_state->buffer, data + copylen, len - copylen);
+    hash_state->buffer_length = len - copylen;
+  }
+}
+
+/**
+ * @brief MD5 typed element processor.
+ *
+ * This accepts typed data, normalizes it, and performs processing on raw bytes.
+ */
+template <typename T>
+void CUDA_DEVICE_CALLABLE md5_process(T const& key, md5_intermediate_data* hash_state)
+{
+  if constexpr (is_fixed_width<T>() && !is_chrono<T>()) {
+    if constexpr (is_floating_point<T>()) {
+      T const normalized_key = normalize_nans_and_zeros_helper<T>(key);
+      uint8_t const* data    = reinterpret_cast<uint8_t const*>(&normalized_key);
+      uint32_t constexpr len = sizeof(T);
+      md5_process_bytes(data, len, hash_state);
+    } else {
+      uint8_t const* data    = reinterpret_cast<uint8_t const*>(&key);
+      uint32_t constexpr len = sizeof(T);
+      md5_process_bytes(data, len, hash_state);
+    }
```

Comment:
Instead of repeating logic in the if/else, just make `normalize_nans_and_zeros` be a passthrough function for non floating point types. 
```suggestion
      T const normalized_key = normalize_nans_and_zeros_helper(key);
      uint8_t const* data    = reinterpret_cast<uint8_t const*>(&normalized_key);
      uint32_t constexpr len = sizeof(T);
      md5_process_bytes(data, len, hash_state);
```